assets/extensions/store.js

```diff
@@ -0,0 +1,292 @@
+/**
+ * External dependencies
+ */
+import { keyBy } from 'lodash';
+
+/**
+ * WordPress dependencies
+ */
+import { registerStore, select, dispatch } from '@wordpress/data';
+import { controls, apiFetch } from '@wordpress/data-controls';
+import { __, sprintf } from '@wordpress/i18n';
+
+const STATUS = {
+	IN_PROGRESS: 'in-progress',
+	IN_QUEUE: 'in-queue',
+};
+
+/**
+ * Checks whether status is a loading status.
+ *
+ * @param {string} status Status to check.
+ *
+ * @return {string} Whether is a loading status.
+ */
+export const isLoadingStatus = ( status ) =>
+	Object.values( STATUS ).includes( status );
+
+/**
+ * Extension store actions.
+ */
+const actions = {
+	/**
+	 * Sets the extensions.
+	 *
+	 * @param {Object}  response                 The response returned from the extensions endpoint.
+	 * @param {Array}   response.extensions      The extensions array.
+	 * @param {boolean} response.wccom_connected True if the site is connected to WC.com.
+	 * @param {boolean} onlyEntities             Whether should update only the extension entities.
+	 */
+	setExtensions(
+		{ extensions, wccom_connected: wcccomConnected },
+		onlyEntities
+	) {
+		const enrichedExtensions = extensions.map( ( extension ) => {
+			let canUpdate = false;
+			// If the extension is hosted in WC.com, check that the site is connected and the subscription is not expired.
+			if ( extension.has_update ) {
+				canUpdate =
+					! extension.wccom_product_id ||
+					( wcccomConnected && ! extension.wccom_expired );
+			}
+
+			return { ...extension, canUpdate };
+		} );
+
+		return {
+			type: 'SET_EXTENSIONS',
+			extensions: enrichedExtensions,
+			onlyEntities,
+		};
+	},
+
+	/**
+	 * Updates the provided extensions.
+	 *
+	 * @param {Array} extensions The extensions to update.
+	 */
+	*updateExtensions( extensions ) {
+		const slugs = extensions.map( ( extension ) => extension.product_slug );
+
+		const inProgressExtensions = yield select(
+			EXTENSIONS_STORE
+		).getExtensionsByStatus( STATUS.IN_PROGRESS );
+
+		// Add extensions to queue and skip if extensions are already updating.
+		if ( inProgressExtensions.length > 0 ) {
+			yield actions.setExtensionsStatus( slugs, STATUS.IN_QUEUE );
```

Comment:
It creates a queue in the frontend to allow the users to click on multiple update buttons in a row. I dug a little into that to see if there was another approach to handle multiple updates simultaneously in the server, but it seems a bit risky to create some issues. We could explore more in the future if we enhance it.

Reply:
The queue code is having some refactor in https://github.com/Automattic/sensei/pull/4214